### DISCUSSION TITLE
I suggest deprecating current implementation of container queries and developing a better one

### DISCUSSION THREAD

**Discussion by varna** [tailwindlabs/tailwindcss#11453] [Ideas]
## The Problem

Tailwind is currently leading all the standards in our industry for CSS. I think that current implementation for container queries is a really bad decision:
1. Using the same naming sense as media queries but providing different values will drive someone insane.
2. Currently provided range of values is not enough. I have to create stuff like `2xs`, `3xs` to make it _usable_ for smaller UI components like sidebars, lists etc.
3. Everyone knows that customising tailwind defaults is a big NO. Just like editing prototype for Array or Object in JS. That's why it is VERY important to provide a proper industry standard for media queries to adopt them properly.
4. Where is support for height?
5. Where is support for max size?
6. [subjective opinion] Using text values instead of `rem` is annoying. I personally mostly use `@[10rem]`, `@[20rem]` and `@[40rem]` and it makes a lot more sense than using `@3xs`, `@xs` and `@1.66xl`. 

## The Solution

1. First step is to gather some smart people with great initial proposals for this API. 
2. Second is to publicize this to no end, so that every dev tweet, blog post and stream would be filled with this.
3. Last step is to collect enough good feedback about certain implementation that we could continue to use it for the next 10 years either it would be desktop, phones, TVs or even VR.

**Comment by varna** [comment]
#### My suggested implementation `@min-w-[value]`, `@min-h-[value]`, `@max-w-[value]`, and `@max-h-[value]`.

- `[value]` - `quarter-rem` used in spacing and sizing or screen name from [screens](https://tailwindcss.com/docs/screens) list.
- `@min-w-80` is the same as `@xs` in current implementation `@container (min-width: 20rem /* 320px */)`
- `@max-h-180` is `@container (max-height: 45rem /* 720px */)`

#### Why?
Self-explanatory. Even a kid can understand what this means, if he knows that `@` is used for container queries.

**Reply by astratagem** [reply]
> Self-explanatory. Even a kid can understand what this means, if he knows that `@` is used for container queries.

I don't think anything is self explanatory when proposing a new syntax / naming convention. The name itself is part of a system of names.

Your suggested implementation, based on my read, breaks from the existing `w-<number>` conventions established by, for example, the [Width utilities](https://tailwindcss.com/docs/width). The `<min|max>-<w|h>-*` utilities do not currently use numeric values, however, so the conflict is indirect. I do think it violates your first assertation about "driving someone insane"…

**Reply by astratagem** [reply]
Also, Tailwind currently discourages `max-*` values in media queries, recommending that they only be used to constrain utilities to a range. Should that apply to container queries?

**Reply by varna** [reply]
Sorry, I made a really big mistake, I actually meant to suggest `quarter-rem`s instead. Like this:

**My suggested implementation `@min-w-[quarter-rem]`, `@min-h-[quarter-rem]`, `@max-w-[quarter-rem]`, and `@max-h-[quarter-rem]`.**
- @min-w-80 is the same as @xs in current implementation `@container (min-width: 20rem /* 320px */)`
- @max-h-180 is `@container (max-height: 45rem /* 720px */)`

> Also, Tailwind currently discourages max-* values in media queries, recommending that they only be used to constrain utilities to a range. Should that apply to container queries?

I know that `max` values are discouraged since the invention of *mobile-first* design.

Personally, I only had to use it once this year because of a very silly feature that my client wanted. I don't even remember what it was, and why I couldn't use `min`, so I can't comment more on this, but I remember seeing another issue with someone asking for it.

I'm suggesting to have word `min` in it to make it more self-explanatory, and I stole the `@` from existing API. I also wanted to steal the spacing and sizing utilities sizes (`quarter-rem`), but I failed and actually wrote down the `rem` sizes instead.

**Comment by astratagem** [comment]
> Currently provided range of values is not enough. I have to create stuff like `2xs`, `3xs` to make it _usable_ for smaller UI components like sidebars, lists etc.

I strongly agree with this. The provided sizes must be scaled down to be of any use. The fact that `xs` begins at `320px` suggests to me that the sizing was initially based on the commonly-encountered media query sizes, which in turn are based on common device viewport widths. But container queries are not media queries.

For practical purposes, I would assert that container widths tend to be narrower than viewport widths, oftentimes as a result of a container whose width is constrained, e.g. by grid columns. When one gets into subdividing the available viewport width, we have to account for the resulting scale reduction.

**Reply by astratagem** [reply]
### Example

https://jsbin.com/kuxopeh/edit?js,console

```js
const viewports = [1024, 1280];
const cols = [3, 4, 6, 8];

const containerWidthsForViewport = (vw) =>
  Object.fromEntries(
    cols.map((c) => [`col-${c}`, Math.round(vw / c)]));

console.log(
  Object.fromEntries(
    viewports.map((w) => [`vw-${w}`, containerWidthsForViewport(w)])
  )
);

```

#### Result

```
{
  vw-1024: {
    col-3: 341,
    col-4: 256,
    col-6: 171,
    col-8: 128
  },
  vw-1280: {
    col-3: 427,
    col-4: 320,
    col-6: 213,
    col-8: 160
  }
}
```

**Comment by mybearworld** [comment]
> * Using the same naming sense as media queries but providing different values will drive someone insane.

This naming convention is used across a lot of different utilities, not just media queries.

> * Everyone knows that customising tailwind defaults is a big NO. Just like editing prototype for Array or Object in JS. That's why it is VERY important to provide a proper industry standard for media queries to adopt them properly.

Not at all - that's why the config itself exists!

> * Where is support for height?
> * Where is support for max size?

Same philosophy as with media queries, use arbitrary values or configuration ([max width](https://tailwindcss.com/docs/screens#max-width-breakpoints), [height](https://tailwindcss.com/docs/screens#custom-media-queries)).

**Reply by varna** [reply]
> This naming convention is used across a lot of different utilities, not just media queries.

Can you provide an example, cause I only know container queries that use `sm` as `@container (min-width: 24rem /* 384px */)`. As much as I remember media queries, containers (`sm:container`) and states (hover, focus, etc.) use `sreens` values instead (i.e.: 'sm': '640px').

So... container queries use sm of 384px, while everyone else uses 640px?

> Not at all - that's why the config itself exists!

Sorry, but no. Before Tailwind was invented, everyone was trying to write their own class names, and it sucked. Currently, Tailwind is THE solution to the problem of having different names and configs for basic things like spacing. If I wanted to write a config, I would have created my own classes for container queries instead.

> Same philosophy as with media queries, use arbitrary values or configuration ([max width](https://tailwindcss.com/docs/screens#max-width-breakpoints), [height](https://tailwindcss.com/docs/screens#custom-media-queries)).

But container queries are used for components and elements instead of screens. If I wanted to make a UI library where you could simply copy some HTML to get a component, like shadcnui, I would need to ship tailwind configs together and that doesn't sound like Tailwind...

<img width="1095" alt="image" src="https://github.com/tailwindlabs/tailwindcss/assets/6717694/6c212bf0-3252-436c-a5d3-8a1c9c6d890b">
